### PR TITLE
Add krel set-release-version command to be used in anago

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This replaces the releaselib implementation for `set_release_version` to
be used from a new `krel set-relese-version` subcommand. The `stdout` of
the command can be used directly to be sourced from anago.

#### Which issue(s) this PR fixes:

None
#### Special notes for your reviewer:
None
#### Does this PR introduce a user-facing change?

```release-note
- Added `krel set-release-version` command to be used from anago
```

### CODE REVIEW

**Comment by saschagrunert** [comment]
Let's give it a shot:
```
> export TOOL_ORG=saschagrunert && export TOOL_BRANCH=krel-set-release-version && go run ./cmd/krel gcbmgr --stage --branch release-1.19 --type official
INFO Verifying repository state
INFO Repository is in clean state
INFO Found matching branch "krel-set-release-version"
INFO Found matching organization "saschagrunert" and repository "release" in remote: origin (github.com:saschagrunert/release)
INFO Verifying remote HEAD commit
INFO Got remote commit: f2e3c939c1498917e5df8ec36149e36b45220a1f
INFO Verifying that remote commit is equal to the local one
INFO Repository is up-to-date
INFO Running gcbmgr with the following options: &{Stage:true Release:false Stream:false Branch:release-1.19 ReleaseType:official BuildVersion: GcpUser: LastJobs:5 Repo:0xc00040dd60 Version:0xc00040dd70}
INFO Build options: {  cloudbuild.yaml   kubernetes-release-test false false false   }
INFO Retrieving Kubernetes release version for ci/latest on branch release-1.19
INFO Retrieving Kubernetes build version from https://dl.k8s.io/ci/latest-1.19.txt...
INFO Retrieved Kubernetes version: v1.19.1-rc.0.34+5f5b46a6e8ad56
INFO Trying to get the kube-cross version for release-1.19...
INFO Found the following kube-cross version: v1.15.0-1
INFO Listing GCB substitutions prior to build submission...
INFO TYPE: official
INFO BUILD_POINT: v1.19.1-rc.0.34-5f5b46a6e8ad56
INFO KUBE_CROSS_VERSION: v1.15.0-1
INFO TOOL_REPO: release
INFO PATCH_VERSION_TAG: 1
INFO KUBERNETES_VERSION_TAG: 1.19.1
INFO GCP_USER_TAG: saschagrunert-at-gmail-com
INFO RELEASE_BRANCH: release-1.19
INFO BUILD_AT_HEAD:
INFO MINOR_VERSION_TAG: 19
INFO NOMOCK_TAG:
INFO TOOL_ORG: saschagrunert
INFO TOOL_BRANCH: krel-set-release-version
INFO TYPE_TAG: official
INFO BUILDVERSION: --buildversion=v1.19.1-rc.0.34+5f5b46a6e8ad56
INFO MAJOR_VERSION_TAG: 1
INFO NOMOCK:
INFO Build directory: /home/sascha/go/src/k8s.io/release/gcb/stage
INFO Config directory: /home/sascha/go/src/k8s.io/release/gcb/stage
INFO Changing to build directory: /home/sascha/go/src/k8s.io/release/gcb/stage
ID                                    CREATE_TIME                DURATION  SOURCE  IMAGES  STATUS
d90222ed-ab75-4376-910c-b50c8ac7e069  2020-09-07T12:26:37+00:00  -         -       -       QUEUED
INFO: Refreshing access_token
Created [https://cloudbuild.googleapis.com/v1/projects/kubernetes-release-test/builds/d90222ed-ab75-4376-910c-b50c8ac7e069].
Logs are available at [https://console.cloud.google.com/cloud-build/builds/d90222ed-ab75-4376-910c-b50c8ac7e069?project=648026197307].
INFO: Display format: "
        table(
          id,
          createTime.date('%Y-%m-%dT%H:%M:%S%Oz', undefined='-'),
          duration(start=startTime,end=finishTime,precision=0,calendar=false,undefined="  -").slice(2:).join(""):label=DURATION,
          build_source(undefined="-"):label=SOURCE,
          build_images(undefined="-"):label=IMAGES,
          status
        )
      "
```

**Comment by saschagrunert** [comment]
Ready for review, PTAL @kubernetes/release-managers

File: cmd/krel/cmd/set_release_version.go
Comment (hasheddan) [inline cmd/krel/cmd/set_release_version.go:29]:
```suggestion
// setReleaseVersionCmd represents the subcommand for `krel set-release-version`
```

File: cmd/krel/cmd/set_release_version.go
Comment (hasheddan) [inline cmd/krel/cmd/set_release_version.go:149]: Can this be a `switch`?

File: pkg/release/release_version_test.go
Comment (hasheddan) [inline pkg/release/release_version_test.go:88]: Quick sanity check for myself here: is this `git.Master` here since we are cutting the release branches later now?

File: lib/releaselib.sh
Comment (saschagrunert) [inline lib/releaselib.sh:54]: :fire:

**Review by hasheddan — COMMENTED** [review]
Thanks for this update @saschagrunert! The subcommand is infinitely easier to understand (for me at least) than `release::set_release_version()`, and we have proper unit tests :) A few comments below :point_down:

File: pkg/release/release_version_test.go
Comment (saschagrunert) [reply pkg/release/release_version_test.go:88]: Yes, I had to actually fix the logic. Luckily it has not used anywhere till now. :)

File: cmd/krel/cmd/set_release_version.go
Comment (saschagrunert) [reply cmd/krel/cmd/set_release_version.go:29]: Fixed :+1:

File: cmd/krel/cmd/set_release_version.go
Comment (saschagrunert) [reply cmd/krel/cmd/set_release_version.go:149]: Hm, since we have different methods to access here I don't think that a switch statement would help us to make it more straight forward to read. How would it look like? :)

File: cmd/krel/cmd/set_release_version.go
Comment (feiskyer) [inline cmd/krel/cmd/set_release_version.go:2]: s/2019/2020/

File: cmd/krel/cmd/set_release_version.go
Comment (feiskyer) [reply cmd/krel/cmd/set_release_version.go:149]: I think switch is not needed here, it is not one variable with different values.

**Review by justaugustus — CHANGES_REQUESTED** [review]
@saschagrunert -- This is amazing!
I've left a series of non-blocking comments. Feel free to remove the hold and address them as follow-ups.

Two things to note:
1. As we're adding more anago logic to krel, does it make sense to make them subcommands instead?
   i.e., `krel set-release-version` becomes `krel anago set-release-version`
2. We've directly translated the bash to Go, but I think we have the opportunity to be more "clever" here, now that we have Go libraries at our disposal.
  
   One potential optimization is taking a release version and being able to extrapolate the next release version:

   ```go
   releaseVersion := <semver>
   releaseVersion.Next() --> <next-semver>
   ```

/lgtm
/approve
/hold

File: cmd/krel/cmd/set_release_version.go
Comment (saschagrunert) [reply cmd/krel/cmd/set_release_version.go]: Thanks for the hint, fixed that! :+1:

File: anago
Comment (saschagrunert) [reply anago:1232]: Okay, I renamed them for now only in our go CLI documentation in `krel gcbmgr` and `krel anago set-release-version`. I can follow-up for anago in another PR. :+1:

File: cmd/krel/cmd/set_release_version.go
Comment (saschagrunert) [reply cmd/krel/cmd/set_release_version.go:26]: Fixed :+1:

File: cmd/krel/cmd/set_release_version_test.go
Comment (saschagrunert) [reply cmd/krel/cmd/set_release_version_test.go:24]: Fixed :+1:

File: pkg/release/release_version.go
Comment (saschagrunert) [reply pkg/release/release_version.go:180]: I think the whole logic inside this function can be simplified and recatored. Before doing that, I'd like to prove that the 1:1 translation from bash to go introduces no regressions. I'll follow-up in a separate PR. :writing_hand:

File: pkg/release/release_version_test.go
Comment (saschagrunert) [reply pkg/release/release_version_test.go:88]: Yes, I can fixup that too in a follow-up PR. :writing_hand:

File: cmd/krel/cmd/set_release_version.go
Comment (saschagrunert) [reply cmd/krel/cmd/set_release_version.go:150]: The main intention for the type with different assessors was to not rely on a map. Maybe this decision makes it more complicated to use the type in the future. I'll revisit and check how we can improve it.

**Comment by saschagrunert** [comment]
Alright, thank you for the feedback y'all!
/hold cancel
@justaugustus I still need to get back my lgtm :innocent:

**Comment by justaugustus** [comment]
SGTM on all of the follow-ups! Great work.
/lgtm
/approve

**Pull request merged**
